File: app/src/main/java/ru/sicampus/bootcamp2025/ui/sign_up/SignUpFragment.java
package ru.sicampus.bootcamp2025.ui.sign_up;

import android.os.Bundle;
import android.text.Editable;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import ru.sicampus.bootcamp2025.R;
import ru.sicampus.bootcamp2025.databinding.FragmentSignUpBinding;
import ru.sicampus.bootcamp2025.ui.utils.Constants;
import ru.sicampus.bootcamp2025.ui.utils.OnChangeText;
import ru.sicampus.bootcamp2025.ui.utils.PreferenceManager;
import ru.sicampus.bootcamp2025.ui.utils.Utils;

public class SignUpFragment extends Fragment {

    private FragmentSignUpBinding binding;
    private PreferenceManager preferenceManager;
    private SignUpViewModel viewModel;
    public SignUpFragment() {
        super(R.layout.fragment_sign_up);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = FragmentSignUpBinding.bind(view);
        preferenceManager = new PreferenceManager(this.getContext());
        viewModel = new ViewModelProvider(this).get(SignUpViewModel.class);
        binding.lastName.addTextChangedListener(new OnChangeText() {
            @Override
            public void afterTextChanged(Editable s) {
                super.afterTextChanged(s);
                viewModel.changeLastName(s.toString());
            }
        });
        binding.firstName.addTextChangedListener(new OnChangeText() {
            @Override
            public void afterTextChanged(Editable s) {
                super.afterTextChanged(s);
                viewModel.changeFirstName(s.toString());
            }
        });
        binding.username.addTextChangedListener(new OnChangeText() {
            @Override
            public void afterTextChanged(Editable s) {
                super.afterTextChanged(s);
                viewModel.changeUsername(s.toString());
            }
        });
        binding.password.addTextChangedListener(new OnChangeText() {
            @Override
            public void afterTextChanged(Editable s) {
                super.afterTextChanged(s);
                viewModel.changePassword(s.toString());
            }
        });
        binding.passwordConfirm.addTextChangedListener(new OnChangeText() {
            @Override
            public void afterTextChanged(Editable s) {
                super.afterTextChanged(s);
                viewModel.changeConfirmPassword(s.toString());
            }
        });
        binding.contactsInfo.addTextChangedListener(new OnChangeText() {
            @Override
            public void afterTextChanged(Editable s) {
                super.afterTextChanged(s);
                viewModel.changeContactInfo(s.toString());
            }
        });
        binding.singUp.setOnClickListener(v -> viewModel.confirm());
        binding.login.setOnClickListener(v -> Navigation.findNavController(view).navigate(R.id.action_signUpFragment_to_signInFragment));
        subscribe(viewModel);
    }

    private void subscribe(SignUpViewModel viewModel) {
        viewModel.errorLiveData.observe(getViewLifecycleOwner(), error -> {
            Toast.makeText(getContext(), error, Toast.LENGTH_SHORT).show();
        });
        /*viewModel.stateLiveData.observe(getViewLifecycleOwner(), state -> {
            binding.confirm.setText(state.getButton());
            binding.title.setText(state.getTitle());
            binding.password.setVisibility(Utils.visibleOrGone(state.isPasswordEnabled()));
        });*/
        viewModel.openListLiveData.observe(getViewLifecycleOwner(), (unused) -> {
            final View view = getView();
            if (view == null) return;
            preferenceManager.putString(Constants.KEY_USER_USERNAME, binding.username.getText().toString());
            view.getRootView().findViewById(R.id.nav_view).setVisibility(View.VISIBLE);
            Navigation.findNavController(view).navigate(R.id.action_signUpFragment_to_mapFragment);
        });
    }

    @Override
    public void onDestroyView() {
        binding = null;
        super.onDestroyView();
    }
}

File: app/src/main/java/ru/sicampus/bootcamp2025/ui/entry/signup/SignUpFragment.kt
package ru.sicampus.bootcamp2025.ui.entry.signup

import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.NavController
import androidx.navigation.fragment.findNavController
import ru.sicampus.bootcamp2025.R
import ru.sicampus.bootcamp2025.databinding.SignupFragmentBinding
import ru.sicampus.bootcamp2025.ui.mainscreen.MainActivity
import ru.sicampus.bootcamp2025.ui.utils.collectWithLifecycle
import ru.sicampus.bootcamp2025.ui.utils.getText
import ru.sicampus.bootcamp2025.ui.utils.visibleOrGone

class SignUpFragment : Fragment(R.layout.signup_fragment) {

    private var _binding: SignupFragmentBinding? = null
    private val binding: SignupFragmentBinding get() = _binding!!

    private var _navController: NavController? = null
    private val navController: NavController get() = _navController!!


    private val viewModel: SignUpViewModel by viewModels<SignUpViewModel> { SignUpViewModel.Factory }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        _binding = SignupFragmentBinding.bind(view)
        _navController = findNavController()

        binding.repeatPassword.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }

            override fun onTextChanged(p0: CharSequence?, p1: Int, p2: Int, p3: Int) {
            }

            override fun afterTextChanged(p0: Editable?) {
                binding.wrapRepeatPassword.error = getText(R.string.password_does_not_matches, requireContext())
                binding.wrapRepeatPassword.isErrorEnabled =
                    binding.repeatPassword.text.toString() != binding.password.text.toString()
            }
        })

        viewModel.state.collectWithLifecycle(this) { state ->
            binding.error.visibility = visibleOrGone(state is SignUpViewModel.State.Error)
            binding.login.isEnabled = state !is SignUpViewModel.State.Loading
            binding.password.isEnabled = state !is SignUpViewModel.State.Loading
            binding.repeatPassword.isEnabled = state !is SignUpViewModel.State.Loading
            when (state) {
                is SignUpViewModel.State.Error -> {
                    binding.error.text = state.errorMessage
                }
                is SignUpViewModel.State.Loading -> Unit
                is SignUpViewModel.State.Waiting -> Unit
            }
        }

        viewModel.action.collectWithLifecycle(this) { action ->
            when (action) {
                SignUpViewModel.Action.OpenApp -> startActivity(
                    Intent(activity, MainActivity::class.java)
                )

                SignUpViewModel.Action.GoToLogin -> toLogin()
            }
        }

        binding.toLogin.setOnClickListener { toLogin() }

        binding.process.setOnClickListener {
            viewModel.onProcessClick(
                binding.login.text.toString(),
                binding.password.text.toString(),
                binding.name.text.toString(),
                binding.lastname.text.toString()
            )
        }
    }

    private fun toLogin() {
        navController.navigate(R.id.login_fragment)
    }

    override fun onDestroyView() {
        _binding = null
        super.onDestroyView()
    }
}